### ISSUE TITLE
Occasionally RuntimeError during training

### ISSUE THREAD

**Issue by mycrazycracy** [phonexiaresearch/VBx-training-recipe#6]
I ran the script and try to reproduce the ResNet101 model. During training, it said:
RuntimeError when processing mini batch 4542/4544. If you see this message frequently, it probably means, that there is some problem with your data or code.
This appeared occasionally. 

I packed the training data using run.sh, and the training procedure seems fine. Could you give me any advice why it appears (and in which condition)? Is this normal and is there anything I should do to fix this?

Part of the training log:
```
[1,0]<stdout>:2021-05-20 11:14:59,509 [train_pytorch_dnn:508 - I] Training will run for 3 epochs = 81 iterations
[1,0]<stdout>:2021-05-20 11:14:59,509 [train_pytorch_dnn:551 - I] Iter: 1/81  Epoch: 0.00/3.0  (0.0% complete)  lr:
0.04000 margin: 0.0500                                                                                              [1,0]<stdout>:2021-05-20 11:18:43,737 [train_pytorch_dnn:247 - I] Train Iter: 1 [9000/81780 (11%)]  Loss: 11.279880
[1,0]<stdout>:2021-05-20 11:22:21,763 [train_pytorch_dnn:247 - I] Train Iter: 1 [18000/81780 (22%)]  Loss: 10.939919
[1,0]<stdout>:2021-05-20 11:26:00,181 [train_pytorch_dnn:247 - I] Train Iter: 1 [27000/81780 (33%)]  Loss: 10.326596
[1,0]<stdout>:2021-05-20 11:29:36,754 [train_pytorch_dnn:247 - I] Train Iter: 1 [36000/81780 (44%)]  Loss: 10.065937[1,0]<stdout>:2021-05-20 11:33:11,309 [train_pytorch_dnn:247 - I] Train Iter: 1 [45000/81780 (55%)]  Loss: 9.550877
[1,0]<stdout>:2021-05-20 11:36:43,507 [train_pytorch_dnn:247 - I] Train Iter: 1 [54000/81780 (66%)]  Loss: 9.324382 [1,0]<stdout>:2021-05-20 11:40:20,921 [train_pytorch_dnn:247 - I] Train Iter: 1 [63000/81780 (77%)]  Loss: 9.245731
[1,0]<stdout>:2021-05-20 11:43:56,940 [train_pytorch_dnn:247 - I] Train Iter: 1 [72000/81780 (88%)]  Loss: 8.924195 [1,0]<stdout>:2021-05-20 11:47:32,667 [train_pytorch_dnn:247 - I] Train Iter: 1 [81000/81780 (99%)]  Loss: 8.405144
[1,1]<stdout>:2021-05-20 11:47:49,679 [train_pytorch_dnn:250 - W] RuntimeError when processing mini batch 4542/4544.
 If you see this message frequently, it probably means, that there is some problem with your data or code.
[1,1]<stdout>:Could not infer dtype of NoneType
[1,0]<stdout>:2021-05-20 11:48:02,697 [train_pytorch_dnn:263 - I] Iteration Loss: 10.5687  Accuracy: 0.00%
[1,0]<stdout>:2021-05-20 11:48:02,698 [train_pytorch_dnn:264 - I] Elapsed time: 33.05 minutes and GPU waiting: 32.94
 minutes.
[1,0]<stdout>:2021-05-20 11:48:02,698 [train_pytorch_dnn:267 - I] Saving model to: exp/nnet/ResNet101_add_margin_emb
ed256_4gpu/models/raw_1.pth
[1,0]<stdout>:2021-05-20 11:48:03,034 [train_pytorch_dnn:551 - I] Iter: 2/81  Epoch: 0.04/3.0  (1.2% complete)  lr:
0.03747 margin: 0.0513
[1,0]<stdout>:2021-05-20 11:51:37,509 [train_pytorch_dnn:247 - I] Train Iter: 2 [9000/81780 (11%)]  Loss: 8.222393
[1,0]<stdout>:2021-05-20 11:55:09,406 [train_pytorch_dnn:247 - I] Train Iter: 2 [18000/81780 (22%)]  Loss: 8.395783
[1,0]<stdout>:2021-05-20 11:58:47,139 [train_pytorch_dnn:247 - I] Train Iter: 2 [27000/81780 (33%)]  Loss: 7.720121
[1,0]<stdout>:2021-05-20 12:02:24,007 [train_pytorch_dnn:247 - I] Train Iter: 2 [36000/81780 (44%)]  Loss: 8.301928
[1,0]<stdout>:2021-05-20 12:06:03,191 [train_pytorch_dnn:247 - I] Train Iter: 2 [45000/81780 (55%)]  Loss: 6.704355
[1,0]<stdout>:2021-05-20 12:09:41,432 [train_pytorch_dnn:247 - I] Train Iter: 2 [54000/81780 (66%)]  Loss: 5.764532
[1,0]<stdout>:2021-05-20 12:13:20,479 [train_pytorch_dnn:247 - I] Train Iter: 2 [63000/81780 (77%)]  Loss: 8.757347
[1,0]<stdout>:2021-05-20 12:17:04,176 [train_pytorch_dnn:247 - I] Train Iter: 2 [72000/81780 (88%)]  Loss: 6.826844
[1,0]<stdout>:2021-05-20 12:20:33,929 [train_pytorch_dnn:247 - I] Train Iter: 2 [81000/81780 (99%)]  Loss: 5.832530
[1,0]<stdout>:2021-05-20 12:21:01,719 [train_pytorch_dnn:263 - I] Iteration Loss: 7.1980  Accuracy: 1.82%
[1,0]<stdout>:2021-05-20 12:21:01,719 [train_pytorch_dnn:264 - I] Elapsed time: 32.98 minutes and GPU waiting: 32.84
 minutes.
[1,0]<stdout>:2021-05-20 12:21:01,720 [train_pytorch_dnn:267 - I] Saving model to: exp/nnet/ResNet101_add_margin_emb
ed256_4gpu/models/raw_2.pth
[1,0]<stdout>:2021-05-20 12:21:02,031 [train_pytorch_dnn:551 - I] Iter: 3/81  Epoch: 0.07/3.0  (2.5% complete)  lr:
0.03509 margin: 0.0526
[1,0]<stdout>:2021-05-20 12:24:40,828 [train_pytorch_dnn:247 - I] Train Iter: 3 [9000/81780 (11%)]  Loss: 5.192419
[1,0]<stdout>:2021-05-20 12:28:12,683 [train_pytorch_dnn:247 - I] Train Iter: 3 [18000/81780 (22%)]  Loss: 4.813242
[1,0]<stdout>:2021-05-20 12:31:45,344 [train_pytorch_dnn:247 - I] Train Iter: 3 [27000/81780 (33%)]  Loss: 5.787135
[1,0]<stdout>:2021-05-20 12:35:18,348 [train_pytorch_dnn:247 - I] Train Iter: 3 [36000/81780 (44%)]  Loss: 6.096598
[1,0]<stdout>:2021-05-20 12:38:50,011 [train_pytorch_dnn:247 - I] Train Iter: 3 [45000/81780 (55%)]  Loss: 4.647009
[1,0]<stdout>:2021-05-20 12:42:21,528 [train_pytorch_dnn:247 - I] Train Iter: 3 [54000/81780 (66%)]  Loss: 4.997063
```

**Comment by Jamiroquai88** [comment]
Hi, honestly, we tried to dig quite deep into python and C++ code and we were not able to find out, what causes this error. In the end, we just added this "stupid" message. If you are able to fix this, we would be grateful.

**Comment by mycrazycracy** [comment]
Thanks for your reply! I just want to make sure whether my training is all right.

After running the training script, I test the model in VoxCeleb1 test set (cos + asnorm is used), and the EER is 1.34%. However, when the released ResNet101 is used, the EER is 0.55%. This is a huge gap. Could you give me some hints what may cause the performance difference? 

The training details:
I use vox1 dev + vox2 dev + cnceleb to train my model and the number of speakers of the training data is exactly 8178.
The model is trained using 4 V100 GPUs and the training took ~2 days (for 3 epoch). This is the training configuration:
horovodrun -np ${ngpus} --mpi --autotune \
    python3 local/train_pytorch_dnn.py --model ${model} \
        --num-targets 8178 \
        --dir ${out_dir}/${model}_${metric}_embed${embed_dim}_${ngpus}gpu \
        --metric ${metric} \
        --egs-dir ${egs_dir} \
        --minibatch-size 18 \
        --embed-dim ${embed_dim} \
        --warmup-epochs 0 \
        --initial-effective-lrate 0.01 \
        --final-effective-lrate 0.00005 \
        --initial-margin-m 0.05 \
        --final-margin-m 0.2 \
        --optimizer SGD \
        --momentum 0.9 \
        --optimizer-weight-decay 0.0001 \
        --preserve-model-interval 30 \
        --num-epochs 3 \
        --apply-cmn no \
        --fix-margin-m 2

The log when the training is over:
![Lark20210602-142945](https://user-images.githubusercontent.com/29195614/120434427-0f667c80-c3af-11eb-9111-2d0e413333c8.png)

**Comment by Jamiroquai88** [comment]
Hey, are you using LDA and other transformations as specified here? https://github.com/phonexiaresearch/VBx-training-recipe/blob/main/local/train_transform.py

**Comment by mycrazycracy** [comment]
No, I just use the cosine distance and AS-Norm to evaluate on VoxCeleb1 test set. No LDA or PLDA is used since cos+asnorm is performed well in speaker verification task (rather than diarization). Any advice would be appreciated.

**Comment by Jamiroquai88** [comment]
Please try to use https://github.com/phonexiaresearch/VBx-training-recipe/blob/main/local/train_transform.py
We could see a huge gain even for speaker verification and diarization too. I would be interested to see if it helped and how much.

**Comment by mycrazycracy** [comment]
I'll try that. Thanks!

And do you have any number of speaker verification (e.g. on vox1) using the ResNet101 model trained by this script? I'll see whether I can reproduce the number. Not sure whether my training data is the same.

**Comment by Jamiroquai88** [comment]
Not sure. I would say it was somewhere below 1% EER.

**Comment by mycrazycracy** [comment]
Thank you for your reply! I'll try the transform you recommend.
